// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeScript Express server generator produces valid code multi-file: endpoint:createUser 1`] = `
"import { CreateUserRequest, CreateUserResponse } from \\"../types\\";

export async function createUser(request: CreateUserRequest, authToken: string | void): Promise<{
    status: 200;
    data: CreateUserResponse;
} | {
    status: number;
    data?: void;
}> {
    throw new Error(\\"Endpoint createUser is not yet implemented!\\");
}"
`;

exports[`TypeScript Express server generator produces valid code multi-file: endpoint:deleteUser 1`] = `
"export async function deleteUser(userId: string, authToken: string): Promise<{
    status: 200;
    data: null;
} | {
    status: 403;
    data: {
        message: string;
        signedInAs: string;
    };
} | {
    status: number;
    data: {
        message: string;
    };
}> {
    throw new Error(\\"Endpoint deleteUser is not yet implemented!\\");
}"
`;

exports[`TypeScript Express server generator produces valid code multi-file: endpoint:getUser 1`] = `
"export async function getUser(userId: string): Promise<{
    status: 200;
    data: {
        name: string;
        age?: number;
    };
} | {
    status: number;
    data?: void;
}> {
    throw new Error(\\"Endpoint getUser is not yet implemented!\\");
}"
`;

exports[`TypeScript Express server generator produces valid code multi-file: server 1`] = `
"import * as cors from \\"cors\\";

import * as express from \\"express\\";

import * as validators from \\"./validators\\";

import { createUser } from \\"./endpoints/createUser\\";

import { deleteUser } from \\"./endpoints/deleteUser\\";

import { getUser } from \\"./endpoints/getUser\\";

const PORT = 3020;

const app = express();

app.use(cors());

app.use(express.json());

app.post(\\"/users/create\\", async (req, res) => {
    const request = req.body;
    if (!validators.validateCreateUser_request(request)) {
        throw new Error(\`Invalid request: \${JSON.stringify(request, null, 2)}\`);
    }
    const authToken = req.headers[\\"Authorization\\"];
    if (!validators.validateCreateUser_headerAuthToken(authToken)) {
        throw new Error(\`Invalid header Authorization: \${JSON.stringify(authToken, null, 2)}\`);
    }
    try {
        const response = await createUser(request, authToken);
        if (response.status >= 200 && response.status < 300) {
            if (!validators.validateCreateUser_response(response.data)) {
                throw new Error(\`Invalid successful response: \${JSON.stringify(response.data, null, 2)}\`);
            }
            res.status(response.status);
            res.json(response.data);
        }
        else {
            res.status(response.status);
            res.end();
        }
    }
    catch (e) {
        console.error(\\"Endpoint createUser threw an unexpected error:\\\\n\\", e);
        res.status(500);
        res.json(\\"An unknown server error has occurred.\\");
    }
});

app.delete(\\"/users/:userId-confirmed\\", async (req, res) => {
    const userId = req.params.userId;
    if (!validators.validateDeleteUser_paramUserId(userId)) {
        throw new Error(\`Invalid path parameter userId: \${JSON.stringify(userId, null, 2)}\`);
    }
    const authToken = req.headers[\\"Authorization\\"];
    if (!validators.validateDeleteUser_headerAuthToken(authToken)) {
        throw new Error(\`Invalid header Authorization: \${JSON.stringify(authToken, null, 2)}\`);
    }
    try {
        const response = await deleteUser(userId, authToken);
        if (response.status === 403) {
            if (!validators.validateDeleteUser_customError403(response.data)) {
                throw new Error(\`Invalid error response for status 403: \${JSON.stringify(response.data, null, 2)}\`);
            }
            res.status(response.status);
            res.json(response.data);
            return;
        }
        if (response.status >= 200 && response.status < 300) {
            if (!validators.validateDeleteUser_response(response.data)) {
                throw new Error(\`Invalid successful response: \${JSON.stringify(response.data, null, 2)}\`);
            }
            res.status(response.status);
            res.json(response.data);
        }
        else {
            if (!validators.validateDeleteUser_genericError(response.data)) {
                throw new Error(\`Invalid error response: \${JSON.stringify(response.data, null, 2)}\`);
            }
            res.status(response.status);
            res.json(response.data);
        }
    }
    catch (e) {
        console.error(\\"Endpoint deleteUser threw an unexpected error:\\\\n\\", e);
        res.status(500);
        res.json(\\"An unknown server error has occurred.\\");
    }
});

app.get(\\"/users/:userId\\", async (req, res) => {
    const userId = req.params.userId;
    if (!validators.validateGetUser_paramUserId(userId)) {
        throw new Error(\`Invalid path parameter userId: \${JSON.stringify(userId, null, 2)}\`);
    }
    try {
        const response = await getUser(userId);
        if (response.status >= 200 && response.status < 300) {
            if (!validators.validateGetUser_response(response.data)) {
                throw new Error(\`Invalid successful response: \${JSON.stringify(response.data, null, 2)}\`);
            }
            res.status(response.status);
            res.json(response.data);
        }
        else {
            res.status(response.status);
            res.end();
        }
    }
    catch (e) {
        console.error(\\"Endpoint getUser threw an unexpected error:\\\\n\\", e);
        res.status(500);
        res.json(\\"An unknown server error has occurred.\\");
    }
});

app.listen(PORT, () => {
    console.log(\`Listening on port \${PORT}\`);
});"
`;

exports[`TypeScript Express server generator produces valid code single-file: endpoint:createUser 1`] = `
"import { CreateUserRequest, CreateUserResponse } from \\"../types\\";

export async function createUser(request: CreateUserRequest, authToken: string | void): Promise<{
    status: 200;
    data: CreateUserResponse;
} | {
    status: number;
    data?: void;
}> {
    throw new Error(\\"Endpoint createUser is not yet implemented!\\");
}"
`;

exports[`TypeScript Express server generator produces valid code single-file: endpoint:deleteUser 1`] = `
"export async function deleteUser(userId: string, authToken: string): Promise<{
    status: 200;
    data: null;
} | {
    status: 403;
    data: {
        message: string;
        signedInAs: string;
    };
} | {
    status: number;
    data: {
        message: string;
    };
}> {
    throw new Error(\\"Endpoint deleteUser is not yet implemented!\\");
}"
`;

exports[`TypeScript Express server generator produces valid code single-file: endpoint:getUser 1`] = `
"export async function getUser(userId: string): Promise<{
    status: 200;
    data: {
        name: string;
        age?: number;
    };
} | {
    status: number;
    data?: void;
}> {
    throw new Error(\\"Endpoint getUser is not yet implemented!\\");
}"
`;

exports[`TypeScript Express server generator produces valid code single-file: server 1`] = `
"import * as cors from \\"cors\\";

import * as express from \\"express\\";

import * as validators from \\"./validators\\";

import { createUser } from \\"./endpoints/createUser\\";

import { getUser } from \\"./endpoints/getUser\\";

import { deleteUser } from \\"./endpoints/deleteUser\\";

const PORT = 3020;

const app = express();

app.use(cors());

app.use(express.json());

app.post(\\"/users/create\\", async (req, res) => {
    const request = req.body;
    if (!validators.validateCreateUser_request(request)) {
        throw new Error(\`Invalid request: \${JSON.stringify(request, null, 2)}\`);
    }
    const authToken = req.headers[\\"Authorization\\"];
    if (!validators.validateCreateUser_headerAuthToken(authToken)) {
        throw new Error(\`Invalid header Authorization: \${JSON.stringify(authToken, null, 2)}\`);
    }
    try {
        const response = await createUser(request, authToken);
        if (response.status >= 200 && response.status < 300) {
            if (!validators.validateCreateUser_response(response.data)) {
                throw new Error(\`Invalid successful response: \${JSON.stringify(response.data, null, 2)}\`);
            }
            res.status(response.status);
            res.json(response.data);
        }
        else {
            res.status(response.status);
            res.end();
        }
    }
    catch (e) {
        console.error(\\"Endpoint createUser threw an unexpected error:\\\\n\\", e);
        res.status(500);
        res.json(\\"An unknown server error has occurred.\\");
    }
});

app.get(\\"/users/:userId\\", async (req, res) => {
    const userId = req.params.userId;
    if (!validators.validateGetUser_paramUserId(userId)) {
        throw new Error(\`Invalid path parameter userId: \${JSON.stringify(userId, null, 2)}\`);
    }
    try {
        const response = await getUser(userId);
        if (response.status >= 200 && response.status < 300) {
            if (!validators.validateGetUser_response(response.data)) {
                throw new Error(\`Invalid successful response: \${JSON.stringify(response.data, null, 2)}\`);
            }
            res.status(response.status);
            res.json(response.data);
        }
        else {
            res.status(response.status);
            res.end();
        }
    }
    catch (e) {
        console.error(\\"Endpoint getUser threw an unexpected error:\\\\n\\", e);
        res.status(500);
        res.json(\\"An unknown server error has occurred.\\");
    }
});

app.delete(\\"/users/:userId-confirmed\\", async (req, res) => {
    const userId = req.params.userId;
    if (!validators.validateDeleteUser_paramUserId(userId)) {
        throw new Error(\`Invalid path parameter userId: \${JSON.stringify(userId, null, 2)}\`);
    }
    const authToken = req.headers[\\"Authorization\\"];
    if (!validators.validateDeleteUser_headerAuthToken(authToken)) {
        throw new Error(\`Invalid header Authorization: \${JSON.stringify(authToken, null, 2)}\`);
    }
    try {
        const response = await deleteUser(userId, authToken);
        if (response.status === 403) {
            if (!validators.validateDeleteUser_customError403(response.data)) {
                throw new Error(\`Invalid error response for status 403: \${JSON.stringify(response.data, null, 2)}\`);
            }
            res.status(response.status);
            res.json(response.data);
            return;
        }
        if (response.status >= 200 && response.status < 300) {
            if (!validators.validateDeleteUser_response(response.data)) {
                throw new Error(\`Invalid successful response: \${JSON.stringify(response.data, null, 2)}\`);
            }
            res.status(response.status);
            res.json(response.data);
        }
        else {
            if (!validators.validateDeleteUser_genericError(response.data)) {
                throw new Error(\`Invalid error response: \${JSON.stringify(response.data, null, 2)}\`);
            }
            res.status(response.status);
            res.json(response.data);
        }
    }
    catch (e) {
        console.error(\\"Endpoint deleteUser threw an unexpected error:\\\\n\\", e);
        res.status(500);
        res.json(\\"An unknown server error has occurred.\\");
    }
});

app.listen(PORT, () => {
    console.log(\`Listening on port \${PORT}\`);
});"
`;
